pkg/api/proxy.go

```diff
@@ -0,0 +1,185 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/dchest/siphash"
+	"github.com/gorilla/mux"
+
+	"zotregistry.dev/zot/pkg/api/config"
+	"zotregistry.dev/zot/pkg/api/constants"
+	"zotregistry.dev/zot/pkg/common"
+)
+
+// ClusterProxy wraps an http.HandlerFunc which requires proxying between zot instances to ensure
+// that a given repository only has a single writer and reader for dist-spec operations in a scale-out cluster.
+// based on the hash value of the repository name, the request will either be handled locally
+// or proxied to another zot member in the cluster to get the data before sending a response to the client.
+func ClusterProxy(ctrlr *Controller) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			config := ctrlr.Config
+			logger := ctrlr.Log
+
+			// if no cluster or single-node cluster, handle locally.
+			if config.Cluster == nil || len(config.Cluster.Members) == 1 {
+				next.ServeHTTP(response, request)
+
+				return
+			}
+
+			// since the handler has been wrapped, it should be possible to get the name
+			// of the repository from the mux.
+			vars := mux.Vars(request)
+			name, ok := vars["name"]
+
+			if !ok || name == "" {
+				response.WriteHeader(http.StatusNotFound)
+
+				return
+			}
+
+			// the target member is the only one which should do read/write for the dist-spec APIs
+			// for the given repository.
+			targetMember := computeTargetMember(config, name)
+
+			// from the member list and our DNS/IP address, calculate if this request should be handled locally.
+			// handle locally, if the "address:port" is a match, else, proxy it to the target member.
+			localMember := fmt.Sprintf("%s:%s", config.HTTP.Address, config.HTTP.Port)
```

Comment:
What if the configured local member IP address is 127.0.0.1 or 0.0.0.0? Maybe we should check if the targetMember is one of the addresses of the local node. Maybe keep a list of the local addresses in a variable.

Reply:
Thanks for the comment @andaaron 
I've addressed this by using the same approach as the sync extension - as suggested.
We now fetch the local IPs from the interfaces as well as resolve hostnames if the member list contains hostnames.

Need to write some Unit Tests for it though.